src/admin_ws.rs

```diff
@@ -0,0 +1,187 @@
+    ///
+    /// This will make multiple attempts according to the `max_retries` and
+    /// `retry_delay_ms` settings, with exponential backoff between retries.
+    pub async fn connect(&mut self) -> HcHttpGatewayResult<()> {
+        self.current_retries = 0;
+
+        while self.current_retries < ADMIN_WS_CONNECTION_MAX_RETRIES {
+            match AdminWebsocket::connect(&self.url).await {
+                Ok(conn) => {
+                    let mut connection = self.connection_handle.lock().map_err(|e| {
+                        HcHttpGatewayError::InternalError(format!(
+                            "Mutex was poisoned during is_connected check: {}",
+                            e
+                        ))
+                    })?;
+
+                    *connection = Some(conn);
+                    self.current_retries = 0;
+
+                    return Ok(());
+                }
+                Err(e) => {
+                    self.current_retries += 1;
+                    tracing::warn!(
+                        "Failed to connect to WebSocket (attempt {}/{}): {:?}",
+                        self.current_retries,
+                        ADMIN_WS_CONNECTION_MAX_RETRIES,
+                        e
+                    );
+
+                    if self.current_retries < ADMIN_WS_CONNECTION_MAX_RETRIES {
+                        sleep(Duration::from_millis(ADMIN_WS_CONNECTION_RETRY_DELAY_MS)).await;
+                    } else {
+                        return Err(HcHttpGatewayError::UpstreamUnavailable);
+                    }
+                }
+            }
+        }
+
+        Err(HcHttpGatewayError::ConfigurationError(format!(
+            "Maximum connection retry attempts ({}) reached",
+            ADMIN_WS_CONNECTION_MAX_RETRIES
+        )))
+    }
+
+    /// Allows calling a method on the AdminWebsocket, with automatic reconnection if needed
+    ///
+    /// Accepts a function `f` that takes a AdminWebsocket and returns a Result with ConductorApiError
+    pub async fn call<T, F, Fut>(&mut self, f: F) -> HcHttpGatewayResult<T>
+    where
+        F: Fn(Arc<AdminWebsocket>) -> Fut + Send + Clone,
+        Fut: Future<Output = ConductorApiResult<T>> + Send,
+    {
+        self.ensure_connected().await?;
+
+        // Try to execute the operation
+        match self.exec(&f).await {
+            Ok(result) => Ok(result),
+            Err(e) if e.is_disconnect_error() => self.handle_disconnection(f).await,
+            Err(e) => Err(e),
+        }
+    }
+
+    // Helper method to execute an operation with the current connection
+    async fn exec<T, F, Fut>(&mut self, f: &F) -> HcHttpGatewayResult<T>
+    where
+        F: Fn(Arc<AdminWebsocket>) -> Fut + Send,
+        Fut: Future<Output = ConductorApiResult<T>> + Send,
+    {
+        // block scope to limit MutexGuard lifetime
+        let connection = {
+            let mut connection = self.connection_handle.lock().map_err(|e| {
```

Comment:
Mutexes can be unwrapped. It's assumed that it's not been poisoned by a panic. If there's a way to panic, I think we better eliminate that than treating this error.